src/fb/nack.rs

```
@@ -0,0 +1,272 @@
+                    if self.base_entry != last_entry {
+                        // we need to output the final entry
+                        let ret = NackBuilderEntryIter::encode_entry(self.base_entry, bitmask);
+                        self.base_entry = last_entry;
+                        return Some(ret);
+                    }
+                }
+                return None;
+            }
+
+            let entry = self.builder.rtp_seq[self.i];
+            if self.i == 0 {
+                // initial set up for the first value
+                self.base_entry = entry;
+                self.i += 1;
+                continue;
+            }
+            let diff = entry.wrapping_sub(self.base_entry);
+            if diff > 0 {
+                bitmask |= 1 << (diff - 1);
+            }
+            if diff >= 16 {
+                // bitmask will overflow next iteration, return the current value
+                let ret = NackBuilderEntryIter::encode_entry(self.base_entry, bitmask);
+                self.base_entry = entry;
+                return Some(ret);
+            }
+            self.i += 1;
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct NackBuilder {
+    rtp_seq: Vec<u16>,
+}
+
+impl NackBuilder {
+    pub fn add_rtp_sequence(mut self, rtp_timestamp: u16) -> Self {
+        self.rtp_seq.push(rtp_timestamp);
+        self.rtp_seq.sort();
+        self
+    }
+
+    fn entries(&self) -> impl Iterator<Item = [u8; 4]> + '_ {
+        NackBuilderEntryIter {
+            builder: self,
+            i: 0,
+            base_entry: 0,
+        }
+    }
+}
+
+impl<'a> FciBuilder<'a> for NackBuilder {
+    fn format(&self) -> u8 {
+        1
+    }
+
+    fn supports_feedback_type(&self) -> FciFeedbackPacketType {
+        FciFeedbackPacketType::TRANSPORT
+    }
+}
+
+impl RtcpPacketWriter for NackBuilder {
+    fn calculate_size(&self) -> Result<usize, RtcpWriteError> {
+        Ok(self.entries().count() * 4)
```

Should we return an error if some entries in `rtp_seq` couldn't be included?